lib/Echidna/Campaign.hs

```diff
@@ -78,10 +81,121 @@ replayCorpus vm txSeqs =
+  -> [(FilePath, [Tx])]
+  -- ^ Initial corpus of transactions
+  -> Maybe Text -- ^ Specified contract name
+  -> [SolcContract] -- ^ List of contracts
+  -> m (WorkerStopReason, WorkerState)
+runSymWorker callback vm dict workerId initialCorpus name cs = do
+  cfg <- asks (.cfg)
+
+  continueLoopMVar <- liftIO $ newMVar () -- used to pause `run` when nothing is happening
+  threadsLeftRef <- liftIO $ newIORef (getNFuzzWorkers cfg.campaignConf)
+  newCovTxs <- liftIO $ newIORef ([] : map snd initialCorpus)
+  txsToApplyToState <- liftIO $ newIORef []
+  let objs = (continueLoopMVar, threadsLeftRef, newCovTxs, txsToApplyToState)
+
+  void $ spawnListener (listenerFunc objs)
+
+  flip runStateT initialState $
+    flip evalRandT (mkStdGen effectiveSeed) $ do -- unused but needed for callseq
+      lift callback
+      void $ replayCorpus vm initialCorpus
+      run objs
+
+  where
+
+  effectiveSeed = dict.defSeed + workerId
+  effectiveGenDict = dict { defSeed = effectiveSeed }
+  initialState =
+    WorkerState { workerId
+                , gasInfo = mempty
+                , genDict = effectiveGenDict
+                , newCoverage = False
+                , ncallseqs = 0
+                , ncalls = 0
+                , runningThreads = []
+                }
+
+  enqueueIORef ref x = atomicModifyIORef' ref (\q -> (q ++ [x], ()))
+
+  dequeueAllIORef ref = atomicModifyIORef' ref ([],)
+
+  dequeueIORef ref = atomicModifyIORef' ref
+    (\case
+      [] -> ([], Nothing)
+      (h:t) -> (t, Just h))
+
+  listenerFunc (continueLoopMVar, _, newCovTxs, txsToApplyToState) (_, WorkerEvent _ (NewCoverage {transactions})) = do
+    enqueueIORef txsToApplyToState transactions
+    enqueueIORef newCovTxs transactions
+    void $ tryPutMVar continueLoopMVar ()
+  listenerFunc (continueLoopMVar, threadsLeftRef, _, _) (_, WorkerEvent _ (WorkerStopped _)) = do
+    newThreadsLeft <- atomicModifyIORef' threadsLeftRef (\n -> (n-1, n-1))
+    when (newThreadsLeft <= 0) $ void $ tryPutMVar continueLoopMVar ()
+  listenerFunc _ _ = pure ()
+
+  run objs@(continueLoopMVar, threadsLeftRef, newCovTxs, txsToApplyToState) = do
+    liftIO $ takeMVar continueLoopMVar
+    mapM_ (callseq vm) =<< liftIO (dequeueAllIORef txsToApplyToState)
+    liftIO (dequeueIORef newCovTxs) >>= \case
+      Nothing -> do
+        -- no more txs for the moment. if everyone else is done, we can finish too
+        threadsLeft <- liftIO (readIORef threadsLeftRef)
+        if threadsLeft <= 0
+          then pure TestLimitReached
+          else continue
+      Just txs -> do
+        symexecTxs txs
+        liftIO $ void $ tryPutMVar continueLoopMVar ()
+        continue
+    where
+    continue = lift callback >> run objs
+
+  symexecTxs txs = do
+    cfg <- asks (.cfg)
+    vm' <- foldlM (\vm' tx -> snd <$> execTx vm' tx) vm txs
+    (threadId, symTxsChan) <- liftIO $ createSymTx cfg name cs vm'
+
+    modify' (\ws -> ws { runningThreads = [threadId] })
```

Comment:
Won't this override other `runningThreads`? Otherwise, why a list here?

Reply:
There's only ever 1 (or 0) runningThreads at a time